lib/api/include/irods/ticketAPI.hpp

```diff
@@ -22,114 +22,132 @@ struct RsComm;
 struct RcComm;
 #endif // IRODS_TICKET_ADMINISTRATION_ENABLE_SERVER_SIDE_API
 
-enum ticket_type
-{
-    READ,
-    WRITE
-};
-
-// WIP: user_constraint, host_constraint, and group_constraint
-enum ticket_prop
-{
-    USES,
-    WRITE_FILE,
-    WRITE_BYTE
-};
-enum ticket_operation
-{
-    SET,
-    REMOVE
-};
-
-class constraint
+/**
+ * @brief Namespace to contain the API for tickets
+ *
+ */
+namespace irods::administration::ticket
 {
-private:
-    ticket_prop property;
-    ticket_operation operation;
-    bool as_admin;
-    int ticket_ID;
-    std::string_view ticket_name;
-
-public:
-    constraint(ticket_prop _prop, ticket_operation _operation, bool admin_access, std::string_view _ticket_name)
-    {
-        property = _prop;
-        operation = _operation;
-        as_admin = admin_access;
-        ticket_name = _ticket_name;
-    }
-    constraint(ticket_prop _prop, ticket_operation _operation, bool admin_access, int ticketID)
+    inline struct admin_tag
     {
-        property = _prop;
-        operation = _operation;
-        as_admin = admin_access;
-        ticket_ID = ticketID;
-    }
-    // constraint()
+    } admin;
 
-    void set_ticket_prop(ticket_prop _prop)
-    {
-        property = _prop;
-    }
-    void set_operation(ticket_operation _oper)
-    {
-        operation = _oper;
-    }
-    void set_admin_access(bool admin_access)
+    enum class ticket_type
     {
-        as_admin = admin_access;
-    }
-    void set_ticket_ID(int ticketID)
+        read,
+        write
+    };
+
+    enum class ticket_property
     {
-        ticket_ID = ticketID;
-    }
-    void set_ticket_name(std::string_view _ticket_name)
+        uses,
+        write_file,
+        write_byte
+    };
+    enum class ticket_operation
     {
-        ticket_name = _ticket_name;
-    }
+        set,
+        remove
+    };
```

Comment:
This is not needed if you take advantage of the type system.

The data types and function name implicitly tell you, as the library implementer, what operation is being performed.

For example, given the following function signature:
```cpp
void add_ticket_constraint(RxComm& _comm,
                           const std::string_view _ticket,
                           const user_constraint& _constraint);
```
I know the following things:
- We are applying a new constraint to a ticket
- We are identifying a ticket by its ticket string (not its integer value)
- We are applying a user constraint

In this fake API function, `user_constraint` simply holds a string that represents an iRODS username. Therefore, this call, if successful, will restrict this ticket's usage to a single user (assuming this ticket didn't have any constraints before invoking the function).

We can go a step further and make this more flexible by turning it into a template like so:
```cpp
template <typename TicketConstraint>
void add_ticket_constraint(RxComm& _comm, 
                           const std::string_view _ticket,
                           const TicketConstraint& _constraint);
```
Now this function can check the deduced type of `TicketConstraint` and delegate to a private function that is responsible for doing the real work.

We do all of this so that the developer only needs to know one function and how to construct a constraint object. The constraint objects can contain any data necessary for doing the operation.

Getter and setter functions don't offer much value here unless the data inside the constraint object is being transformed in some way or we expect that can happen in the future. I believe we can safely infer that that won't happen because iRODS tickets already have a clear definition, regardless of how they are implemented today.

Reply:
@secretstf ping